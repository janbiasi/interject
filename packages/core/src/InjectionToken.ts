import { ProviderScope } from './Scope';
import { IInjectable } from './Injectable';

interface InjectionTokenOptions<T> {
	scope: ProviderScope;
	factory: () => T;
}

const INJECTION_TOKEN_DEFAULT_OPTIONS = {
	scope: 'root',
	factory: () => undefined,
};

export class InjectionToken<T> {
	readonly metadataName = 'InjectionToken';
	readonly injectable: IInjectable<T>;

	constructor(private readonly _id: string, options: InjectionTokenOptions<T> = INJECTION_TOKEN_DEFAULT_OPTIONS) {
		this.injectable = {
			scope: options.scope,
			factory: options.factory,
			value: undefined,
		};
	}

	toString() {
		const scope = this.injectable.scope;
		const displayPossibleScope = scope && scope !== 'root' ? `@${scope}` : '';

		return `InjectionToken${displayPossibleScope} ${this._id}`;
	}
}
