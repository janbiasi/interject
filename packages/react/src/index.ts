export { withInterject } from './withInterject';
export {Â ReactProvidableModule } from './utils';