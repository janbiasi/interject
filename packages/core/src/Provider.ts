import { IConstructable } from './Constructable';
import { InjectionToken } from './InjectionToken';

export type Provide<T = any> = InjectionToken<T> | IConstructable<any>;

export interface IValueProvider {
	useValue: any;
	provide: Provide;
	multi?: boolean;
}

export interface IClassProvider {
	useClass: IConstructable<any>;
	provide: Provide;
	requires?: any[];
}

export interface IExistingProvider {
	useExisting: any;
	provide: Provide;
}

export interface IFactoryProvider {
	useFactory: Function;
	provide: Provide;
	requires?: any[];
}

export type Provider = IValueProvider | IExistingProvider | IClassProvider | IFactoryProvider;
