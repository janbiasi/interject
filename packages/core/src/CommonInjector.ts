import { IConstructable } from './Constructable';
import { Provider } from './Provider';
import { THROW_IF_NOT_FOUND } from './utils';
import { InjectionToken } from './InjectionToken';
import { InjcetionFlags } from './Flags';
import { IInjectable } from './Injectable';
import { Injector } from './Injector';

// runtime
import { setCurrentInjector, getCurrentInjector } from './runtime';

// core providers
import VersionProvider from './platform/providers/Version';

export interface IInjectorType<T> extends IConstructable<T> {
	injector: IInjector<T>;
}

export interface IInjectorTypeWithProviders<T> {
	pointer: IInjectorType<T>;
	providers?: Provider[];
}

export interface IInjector<T> {
	factory: () => T;
	providers: Provider[];
	imports: (IInjectorType<any> | IInjectorTypeWithProviders<any>)[];
}

export interface ICommonInjectorCreateOptions {
	providers: Provider[];
	parent?: CommonInjector;
	name?: string;
}

export class NullInjector implements CommonInjector {
	get(token: any, notFoundValue: any = THROW_IF_NOT_FOUND): any {
		if (notFoundValue === THROW_IF_NOT_FOUND) {
			throw new Error(`NullInjectorError: No provider for ${token}!`);
		}

		return notFoundValue;
	}
}

export abstract class CommonInjector {
	static NULL: CommonInjector = new NullInjector();
	static THROW_IF_NOT_FOUND = THROW_IF_NOT_FOUND;

	abstract get<T>(token: IConstructable<T> | InjectionToken<T>, notFoundValue?: T, flags?: InjcetionFlags): T;

	static create(options: ICommonInjectorCreateOptions): CommonInjector {
		// add core providers
		options.providers.push(VersionProvider);

		return new Injector(options.providers, options.parent, options.name || null);
	}

	static injectable = <IInjectable<any>>{
		providedIn: 'any',
		factory: () => undefined as any,
		scope: 'root',
		value: undefined,
	};

	static toString() {
		return `CommonInjector`;
	}

	static get currentInjector(): CommonInjector {
		return getCurrentInjector();
	}
}
