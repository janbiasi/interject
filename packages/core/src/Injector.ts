import { CommonInjector } from './CommonInjector';
import { Provider } from './Provider';
import { IInjection } from './InjectionItem';
import { EMPTY, CIRCULAR, stringify } from './utils';
import { InjectionToken } from './InjectionToken';
import { IConstructable } from './Constructable';
import { InjcetionFlags } from './Flags';
import { resolveToken, resolveInjectableProvider } from './resolve';
import { setCurrentInjector } from './runtime';

export class Injector implements CommonInjector {
	readonly parent: CommonInjector;
	readonly source: string | null;

	private _injections: Map<any, IInjection>;

	constructor(providers: Provider[], parent: CommonInjector = CommonInjector.NULL, source: string | null = null) {
		this.parent = parent;
		this.source = source;

		// enable runtime features and closure scoping
		setCurrentInjector(this);

		// storage for all injectables
		const injections = (this._injections = new Map<any, IInjection>());

		// set an injection point to the common injector
		injections.set(CommonInjector, <IInjection>{
			token: CommonInjector,
			factory: CIRCULAR,
			dependencies: EMPTY,
			value: this,
			useNew: false,
		});

		// process all providers
		providers.forEach((provider: Provider) => resolveInjectableProvider(provider, injections));
	}

	set(token: InjectionToken<any> | IConstructable<any>, injection: IInjection): void {
		this._injections.set(token, injection);
	}

	get<T>(token: InjectionToken<any> | IConstructable<any>, notFoundValue?: T, flags?: InjcetionFlags): T {
		const injection = this._injections.get(token);

		// resolve the token and its dependencies
		return resolveToken(token, injection, this._injections, this.parent, notFoundValue, flags);
	}

	toString() {
		const tokens: string[] = [];
		this._injections.forEach((key, token) => tokens.push(stringify(token)));

		return `Injector {Â ${tokens.join(', ')} }`;
	}
}
