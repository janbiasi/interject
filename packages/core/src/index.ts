import { CommonInjector } from './CommonInjector';
import { Injector } from './Injector';
import { InjectionToken } from './InjectionToken';
import { InjcetionFlags, LookupFlags } from './Flags';
import { Provider, IValueProvider, IExistingProvider, IClassProvider, IFactoryProvider } from './Provider';
import { Injectable, Module, internalModuleDecoratorFactory } from './decorators';

export {
	// basic core
	Injector,
	CommonInjector,
	InjectionToken,
	// flags
	InjcetionFlags,
	LookupFlags,
	// providers
	Provider,
	IValueProvider,
	IExistingProvider,
	IClassProvider,
	IFactoryProvider,
	// internal decorator factories
	internalModuleDecoratorFactory,
	// decorators
	Injectable,
	Module,
};
