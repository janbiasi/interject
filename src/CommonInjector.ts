import { IConstructable } from './Constructable';
import { Provider } from './Provider';
import { THROW_IF_NOT_FOUND, CONCAT, EMPTY } from './utils';
import { InjectionToken } from './InjectionToken';
import { LookupFlags, InjcetionFlags } from './Flags';
import { IInjectable } from './Injectable';
import { Injector } from './Injector';
import { IInjection } from './InjectionItem';

export interface IInjectorType<T> extends IConstructable<T> {
	injector: IInjector<T>;
}

export interface IInjectorTypeWithProviders<T> {
	pointer: IInjectorType<T>;
	providers?: Provider[];
}

export interface IInjector<T> {
	factory: () => T;
	providers: Provider[];
	imports: (IInjectorType<any> | IInjectorTypeWithProviders<any>)[];
}

export type CommonInjectorCreateOptions =
	| Provider[]
	| {
			providers: Provider[];
			parent?: CommonInjector;
			name?: string;
	  };

export class NullInjector implements CommonInjector {
	get(token: any, notFoundValue: any = THROW_IF_NOT_FOUND): any {
		if (notFoundValue === THROW_IF_NOT_FOUND) {
			throw new Error(`NullInjectorError: No provider for ${token}!`);
		}

		return notFoundValue;
	}
}

export abstract class CommonInjector {
	static NULL: CommonInjector = new NullInjector();
	static THROW_IF_NOT_FOUND = THROW_IF_NOT_FOUND;

	abstract get<T>(token: IConstructable<T> | InjectionToken<T>, notFoundValue?: T, flags?: InjcetionFlags): T;

	static create(options: CommonInjectorCreateOptions): CommonInjector;
	static create(options: CommonInjectorCreateOptions, parent?: CommonInjector): CommonInjector {
		if (Array.isArray(options)) {
			return new Injector(options, parent);
		} else {
			return new Injector(options.providers, options.parent, options.name || null);
		}
	}

	static injectable = <IInjectable<any>>{
		providedIn: 'any',
		factory: () => undefined as any,
		scope: 'root',
		value: undefined,
	};
}

export const resolve = (
	token: any,
	injection: IInjection | undefined,
	injections: Map<any, IInjection>,
	parent: CommonInjector,
	notFoundValue: any,
	flag: InjcetionFlags
): any => {
	let value: any = injection.value || undefined;

	if (flag === InjcetionFlags.Skip) {
		return parent.get(token, notFoundValue, InjcetionFlags.Default);
	}

	if (injection.useNew && injection.value === EMPTY) {
		injection.value = value = new (injection.factory as any)(...injection.dependencies);
	} else if (injection.useNew && injection.value !== EMPTY) {
		// do nothing, class already instanciated
	} else if (typeof injection.factory === 'function' && injection.factory !== CONCAT) {
		// apply deps to factory and set it to the value
		value = injection.value = injection.factory.apply(null, injection.dependencies);
	}

	return value;
};
