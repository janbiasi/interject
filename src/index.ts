import { Injector } from './Injector';
import { InjectionToken } from './InjectionToken';
import { InjcetionFlags, LookupFlags } from './Flags';
import { Provider } from './Provider';
import {Â ProviderScope } from './Scope';

export { Injector, InjectionToken, InjcetionFlags, LookupFlags, Provider, ProviderScope };
