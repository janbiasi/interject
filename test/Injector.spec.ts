import { Injector } from '../src/Injector';
import { InjectionToken } from '../src/InjectionToken';
import { CommonInjector } from '../src/CommonInjector';
import { IClassProvider, IValueProvider } from 'src/Provider';

beforeEach(() => {
	jest.resetModules();
});

describe('Injector', () => {
	describe('basics', () => {
		it('should be creatable without providers', () => {
			const injector = new Injector([]);
			expect(injector).toMatchSnapshot();
		});

		it('should be creatable with providers', () => {
			const testProviderToken = new InjectionToken('withProvidersToken');
			const injector = new Injector([
				{
					provide: testProviderToken,
					useValue: 10,
				},
			]);

			expect(injector).toMatchSnapshot();
		});
	});

	describe('inheritance', () => {
		it('should resolve from parent', () => {
			const testProviderToken = new InjectionToken<number>('inheritProviderToken');
			const parentInjector = new Injector([
				{
					provide: testProviderToken,
					useValue: 10,
				},
			]);
			const childInjector = new Injector([], parentInjector);
			const getServiceFromParentThruChild = childInjector.get(testProviderToken);

			expect(childInjector).toMatchSnapshot();
			expect(childInjector.toString()).toMatchSnapshot();
			expect(getServiceFromParentThruChild).toBeDefined();
			expect(getServiceFromParentThruChild).toEqual(10);
		});
	});

	describe('specific providers', () => {
		describe('ValueProvider', () => {
			it('should resolve a single value from a value provider', () => {
				const testProviderToken = new InjectionToken<number>('valueProviderToken');
				const testProvider = {
					provide: testProviderToken,
					useValue: 10,
				};
				const injector = new Injector([testProvider]);

				expect(injector).toMatchSnapshot();
				expect(injector.toString()).toMatchSnapshot();
				expect(injector.get(testProviderToken)).toEqual(10);
			});

			it('should resolve multiple values from a value provider', () => {
				const testProviderToken = new InjectionToken<number>('multiProviderToken');
				const injector = new Injector([
					{
						provide: testProviderToken,
						useValue: 10,
					},
					{
						provide: testProviderToken,
						useValue: 20,
					},
				]);

				expect(injector).toMatchSnapshot();
				expect(injector.toString()).toMatchSnapshot();
				expect(injector.get(testProviderToken)).toEqual([10, 20]);
			});
		});

		describe('ClassProvider', () => {
			it('should resolve a single class from a provider', () => {
				class TestService {
					static instanceCount = 0;

					constructor() {
						TestService.instanceCount++;
					}

					public get value() {
						return 'some value';
					}
				}

				const testServiceToken = new InjectionToken<TestService>('classProviderToken');
				const injector = new Injector([
					{
						provide: testServiceToken,
						useClass: TestService,
					},
				]);
				const injectedService = injector.get(testServiceToken);

				expect(injector).toMatchSnapshot();
				expect(injector.toString()).toMatchSnapshot();
				expect(injectedService.value).toBeDefined();
				expect(injectedService.value).toEqual('some value');
			});

			it('should resolve multiple request to the same instance', () => {
				class TestService {
					static instanceCount = 0;

					constructor() {
						TestService.instanceCount++;
					}

					public get value() {
						return 'some value';
					}
				}

				const testServiceToken = new InjectionToken<TestService>('singletonClassProviderToken');
				const injector = new Injector([
					{
						provide: testServiceToken,
						useClass: TestService,
					},
				]);
				const injectedService = injector.get(testServiceToken);
				const injectedServiceClone1 = injector.get(testServiceToken);
				const injectedServiceClone2 = injector.get(testServiceToken);

				expect(injector).toMatchSnapshot();
				expect(injector.toString()).toMatchSnapshot();
				expect(injectedService.value).toBeDefined();
				expect(injectedService.value).toEqual('some value');
				expect(injectedService).toMatchObject(injectedServiceClone1);
				expect(injectedService).toMatchObject(injectedServiceClone2);
				expect(TestService.instanceCount).toEqual(1);
			});

			it.only('should resolve required dependencies', () => {
				const testServiceToken = new InjectionToken<TestService>('testServiceToken');
				const valueServiceToken = new InjectionToken<ValueService>('valueServiceToken');
				const configToken = new InjectionToken<{env: string}>('configToken');

				class ValueService {
					constructor(private config: {env: string}) {}

					public get value() {
						return this.config.env;
					}
				}

				class TestService {
					constructor(private valueService: ValueService) {}

					public get value() {
						return this.valueService.value;
					}
				}

				const injector = CommonInjector.create({
					providers: [
						<IValueProvider>{
							provide: configToken,
							useValue: { env: 'development' }
						},
						<IClassProvider>{
							provide: valueServiceToken,
							useClass: ValueService,
							requires: [configToken]
						},
						<IClassProvider>{
							provide: testServiceToken,
							useClass: TestService,
							requires: [valueServiceToken],
						},
					],
				});

				expect(injector).toMatchSnapshot();
				expect(injector.get(testServiceToken).value).toEqual('development');
			});
		});

		describe('FactoryProvider', () => {
			it('should resolve the factory value correctly', () => {
				const testFactoryToken = new InjectionToken<{ value: string }>('factoryProviderToken');
				const testFactory = () => ({
					value: ['value'],
				});
				const testFactoryProvider = {
					provide: testFactoryToken,
					useFactory: testFactory,
				};
				const injector = new Injector([testFactoryProvider]);
				const result = injector.get(testFactoryToken);

				expect(injector).toMatchSnapshot();
				expect(result.value).toEqual(['value']);
			});

			it('should execute the factory only once', () => {
				const testFactoryToken = new InjectionToken<{ value: string }>('multiFactoryProviderToken');
				let factoryCreationCounter = 0;
				const testFactory = () => {
					factoryCreationCounter++;
					return { value: ['value'] };
				};
				const testFactoryProvider = {
					provide: testFactoryToken,
					useFactory: testFactory,
				};
				const injector = new Injector([testFactoryProvider]);
				const firstGet = injector.get(testFactoryToken);
				const secondGet = injector.get(testFactoryToken);
				const thirdGet = injector.get(testFactoryToken);

				expect(injector).toMatchSnapshot();
				expect(firstGet.value).toEqual(['value']);
				expect(secondGet.value).toEqual(['value']);
				expect(thirdGet.value).toEqual(['value']);
				expect(factoryCreationCounter).toEqual(1);
			});
		});
	});
});
