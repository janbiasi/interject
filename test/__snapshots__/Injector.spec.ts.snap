// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Injector basics should be creatable with providers 1`] = `
Injector {
  "_injections": Map {
    [Function] => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": [Function],
      "useNew": false,
      "value": [Circular],
    },
    InjectionToken {
      "_id": "withProvidersToken",
      "injectable": Object {
        "factory": [Function],
        "scope": "root",
        "value": undefined,
      },
      "metadataName": "InjectionToken",
    } => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": InjectionToken {
        "_id": "withProvidersToken",
        "injectable": Object {
          "factory": [Function],
          "scope": "root",
          "value": undefined,
        },
        "metadataName": "InjectionToken",
      },
      "useNew": false,
      "value": 10,
    },
  },
  "parent": NullInjector {},
  "source": null,
}
`;

exports[`Injector basics should be creatable without providers 1`] = `
Injector {
  "_injections": Map {
    [Function] => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": [Function],
      "useNew": false,
      "value": [Circular],
    },
  },
  "parent": NullInjector {},
  "source": null,
}
`;

exports[`Injector inheritance should resolve from parent 1`] = `
Injector {
  "_injections": Map {
    [Function] => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": [Function],
      "useNew": false,
      "value": [Circular],
    },
  },
  "parent": Injector {
    "_injections": Map {
      [Function] => Object {
        "dependencies": Array [],
        "factory": [Function],
        "token": [Function],
        "useNew": false,
        "value": [Circular],
      },
      InjectionToken {
        "_id": "inheritProviderToken",
        "injectable": Object {
          "factory": [Function],
          "scope": "root",
          "value": undefined,
        },
        "metadataName": "InjectionToken",
      } => Object {
        "dependencies": Array [],
        "factory": [Function],
        "token": InjectionToken {
          "_id": "inheritProviderToken",
          "injectable": Object {
            "factory": [Function],
            "scope": "root",
            "value": undefined,
          },
          "metadataName": "InjectionToken",
        },
        "useNew": false,
        "value": 10,
      },
    },
    "parent": NullInjector {},
    "source": null,
  },
  "source": null,
}
`;

exports[`Injector inheritance should resolve from parent 2`] = `"Injector { CommonInjector }"`;

exports[`Injector specific providers ClassProvider should resolve a single class from a provider 1`] = `
Injector {
  "_injections": Map {
    [Function] => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": [Function],
      "useNew": false,
      "value": [Circular],
    },
    InjectionToken {
      "_id": "classProviderToken",
      "injectable": Object {
        "factory": [Function],
        "scope": "root",
        "value": undefined,
      },
      "metadataName": "InjectionToken",
    } => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": InjectionToken {
        "_id": "classProviderToken",
        "injectable": Object {
          "factory": [Function],
          "scope": "root",
          "value": undefined,
        },
        "metadataName": "InjectionToken",
      },
      "useNew": true,
      "value": TestService {},
    },
  },
  "parent": NullInjector {},
  "source": null,
}
`;

exports[`Injector specific providers ClassProvider should resolve a single class from a provider 2`] = `"Injector { CommonInjector, InjectionToken classProviderToken }"`;

exports[`Injector specific providers ClassProvider should resolve multiple request to the same instance 1`] = `
Injector {
  "_injections": Map {
    [Function] => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": [Function],
      "useNew": false,
      "value": [Circular],
    },
    InjectionToken {
      "_id": "singletonClassProviderToken",
      "injectable": Object {
        "factory": [Function],
        "scope": "root",
        "value": undefined,
      },
      "metadataName": "InjectionToken",
    } => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": InjectionToken {
        "_id": "singletonClassProviderToken",
        "injectable": Object {
          "factory": [Function],
          "scope": "root",
          "value": undefined,
        },
        "metadataName": "InjectionToken",
      },
      "useNew": true,
      "value": TestService {},
    },
  },
  "parent": NullInjector {},
  "source": null,
}
`;

exports[`Injector specific providers ClassProvider should resolve multiple request to the same instance 2`] = `"Injector { CommonInjector, InjectionToken singletonClassProviderToken }"`;

exports[`Injector specific providers FactoryProvider should execute the factory only once 1`] = `
Injector {
  "_injections": Map {
    [Function] => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": [Function],
      "useNew": false,
      "value": [Circular],
    },
    InjectionToken {
      "_id": "multiFactoryProviderToken",
      "injectable": Object {
        "factory": [Function],
        "scope": "root",
        "value": undefined,
      },
      "metadataName": "InjectionToken",
    } => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": InjectionToken {
        "_id": "multiFactoryProviderToken",
        "injectable": Object {
          "factory": [Function],
          "scope": "root",
          "value": undefined,
        },
        "metadataName": "InjectionToken",
      },
      "useNew": false,
      "value": Object {
        "value": Array [
          "value",
        ],
      },
    },
  },
  "parent": NullInjector {},
  "source": null,
}
`;

exports[`Injector specific providers FactoryProvider should resolve the factory value correctly 1`] = `
Injector {
  "_injections": Map {
    [Function] => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": [Function],
      "useNew": false,
      "value": [Circular],
    },
    InjectionToken {
      "_id": "factoryProviderToken",
      "injectable": Object {
        "factory": [Function],
        "scope": "root",
        "value": undefined,
      },
      "metadataName": "InjectionToken",
    } => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": InjectionToken {
        "_id": "factoryProviderToken",
        "injectable": Object {
          "factory": [Function],
          "scope": "root",
          "value": undefined,
        },
        "metadataName": "InjectionToken",
      },
      "useNew": false,
      "value": Object {
        "value": Array [
          "value",
        ],
      },
    },
  },
  "parent": NullInjector {},
  "source": null,
}
`;

exports[`Injector specific providers ValueProvider should resolve a single value from a value provider 1`] = `
Injector {
  "_injections": Map {
    [Function] => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": [Function],
      "useNew": false,
      "value": [Circular],
    },
    InjectionToken {
      "_id": "valueProviderToken",
      "injectable": Object {
        "factory": [Function],
        "scope": "root",
        "value": undefined,
      },
      "metadataName": "InjectionToken",
    } => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": InjectionToken {
        "_id": "valueProviderToken",
        "injectable": Object {
          "factory": [Function],
          "scope": "root",
          "value": undefined,
        },
        "metadataName": "InjectionToken",
      },
      "useNew": false,
      "value": 10,
    },
  },
  "parent": NullInjector {},
  "source": null,
}
`;

exports[`Injector specific providers ValueProvider should resolve a single value from a value provider 2`] = `"Injector { CommonInjector, InjectionToken valueProviderToken }"`;

exports[`Injector specific providers ValueProvider should resolve multiple values from a value provider 1`] = `
Injector {
  "_injections": Map {
    [Function] => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": [Function],
      "useNew": false,
      "value": [Circular],
    },
    InjectionToken {
      "_id": "multiProviderToken",
      "injectable": Object {
        "factory": [Function],
        "scope": "root",
        "value": undefined,
      },
      "metadataName": "InjectionToken",
    } => Object {
      "dependencies": Array [],
      "factory": undefined,
      "token": InjectionToken {
        "_id": "multiProviderToken",
        "injectable": Object {
          "factory": [Function],
          "scope": "root",
          "value": undefined,
        },
        "metadataName": "InjectionToken",
      },
      "useNew": false,
      "value": Array [
        10,
        20,
      ],
    },
  },
  "parent": NullInjector {},
  "source": null,
}
`;

exports[`Injector specific providers ValueProvider should resolve multiple values from a value provider 2`] = `"Injector { CommonInjector, InjectionToken multiProviderToken }"`;
